generator/src/main/java/com/scottlogic/deg/generator/generation/grouped/FieldSpecGroupValueGenerator.java

```diff
@@ -17,172 +17,177 @@
 
 public class FieldSpecGroupValueGenerator {
 
+    private final CombinationStrategyType combinationStrategy;
     private final FieldSpecValueGenerator underlyingGenerator;
+    private final FieldSpecMerger fieldSpecMerger = new FieldSpecMerger();
 
-    public FieldSpecGroupValueGenerator(FieldSpecValueGenerator underlyingGenerator) {
+    @Inject
+    public FieldSpecGroupValueGenerator(FieldSpecValueGenerator underlyingGenerator, CombinationStrategyType combinationStrategy) {
         this.underlyingGenerator = underlyingGenerator;
+        this.combinationStrategy = combinationStrategy;
     }
 
     public Stream<DataBag> generate(FieldSpecGroup group) {
-        Field first = SetUtils.firstIteratorElement(group.fieldSpecs().keySet());
+        Field first = getFirst(group);
 
-        FieldSpecGroup groupRespectingFirstField = initialAdjustments(first, group);
-        FieldSpec firstSpec = groupRespectingFirstField.fieldSpecs().get(first);
+        if (group.fieldSpecs().size() == 1){
+            return underlyingGenerator.generate(first, group.fieldSpecs().get(first))
+                .map(val -> toDataBag(first, val));
+        }
+
+        FieldSpec firstSpec = updateFirstSpecFromRelations(first, group);
 
         Stream<DataBag> firstDataBagValues = underlyingGenerator.generate(first, firstSpec)
             .map(value -> toDataBag(first, value));
 
-        return createRemainingDataBags(firstDataBagValues, first, groupRespectingFirstField);
+        return flatMap(
+            firstDataBagValues,
+            dataBag -> generateRemainingData(first, dataBag, removeSpecFromGroup(first, group)));
     }
 
-    private static DataBag toDataBag(Field field, DataBagValue value) {
-        Map<Field, DataBagValue> map = new HashMap<>();
-        map.put(field, value);
-        return new DataBag(map);
+    private Field getFirst(FieldSpecGroup keySet) {
+        Map<Field, Integer> otherCount = keySet.relations().stream().collect(Collectors.toMap(
+            r -> r.other(),
+            r -> 1, (l, r) -> l + r));
+        return otherCount.keySet().stream().reduce((l,r)->otherCount.get(l)>otherCount.get(r)?l:r)
+            .orElse(keySet.fieldSpecs().keySet().iterator().next());
     }
 
-    private static FieldSpecGroup initialAdjustments(Field first, FieldSpecGroup group) {
-        checkOnlyPairwiseRelationsExist(group.relations());
-
-        Map<Field, FieldSpec> mutatingSpecs = new HashMap<>(group.fieldSpecs());
-
-        for (FieldSpecRelations relation : group.relations()) {
-            FieldSpec merged = createMergedSpecFromRelation(first, relation, group)
-                .orElseThrow(() -> new IllegalStateException("Failed to merge field specs in related fields"));
-            mutatingSpecs.replace(first, merged);
-        }
+    private FieldSpec updateFirstSpecFromRelations(Field first, FieldSpecGroup group) {
+        Optional<FieldSpec> optionalFirstFieldSpec = Optional.of(group.fieldSpecs().get(first));
```

Comment:
should get(first) always return non-null?  if so we cna get rid of this Optional wrapping that doing which is complictaing things.  If not is their anything that cna do to get rid of tehse Optionals as it complicates code.  Also mergeOptionlFieldSpecs always assumes right param is present.  Therefore its only ever this get(first) result possibly returning null that creates the need for this Optional stuff

Reply:
The only reason it's optional is so it works with the merger in the reduce function.

i've shifted it about a bit so this is less unclear